### PR TITLE
Update pars

### PR DESCRIPTION
**Pull request by hillalex**
Allow updating model parameters

### CODE REVIEW

**Review by hillalex — COMMENTED** [review]
Do these tests seem to cover the desired behaviour? I assume time would generally be reset along with parameters, but leaving that up to the user to specify

File: inst/include/mode/r/helpers.hpp [hunk 1]
```diff
@@ -117,12 +117,26 @@ std::vector<double> validate_state(cpp11::sexp r_state,
...
+}
+
+inline
+bool validate_set_initial_state(SEXP r_state, SEXP r_pars, SEXP r_time,
                                 SEXP r_set_initial_state) {
   bool set_initial_state = false;
   if (r_set_initial_state == R_NilValue) {
     set_initial_state = r_state == R_NilValue &&
-                        r_time != R_NilValue;
+        (r_time != R_NilValue || r_pars != R_NilValue);
```
Comment (richfitz) [inline inst/include/mode/r/helpers.hpp:137]: the boolean in dust is 

```
r_state == R_NilValue && r_pars != R_NilValue && r_step != R_NilValue
```

but here is

```
r_state == R_NilValue && (r_time != R_NilValue || r_pars != R_NilValue)
```

which is slightly different. I think that the dust one is what we want - if time has not been reset we don't want to set the initial state. Also happy to park this for later and potentially change the dust behaviour if we think that's less consistent

File: inst/include/mode/r/helpers.hpp [hunk 2]
```diff
@@ -117,12 +117,26 @@ std::vector<double> validate_state(cpp11::sexp r_state,
...
+  if (r_time == R_NilValue) {
+    return std::vector<double>(0);
+  }
+  cpp11::doubles time = cpp11::as_cpp<cpp11::doubles>(r_time);
+  if (time.size() != 1) {
+    cpp11::stop("Expected 'time' to be a scalar value");
+  }
+  std::vector<double> ret(1);
+  std::copy_n(REAL(time.data()), 1, ret.begin());
+  return ret;
```
Comment (richfitz) [inline inst/include/mode/r/helpers.hpp]: consider:

```suggestion
  return std::vector<double> {time[0]};
```

**Review by richfitz — COMMENTED** [review]
Tests look fine for now - doesn't look like there are any other good ones to port over from dust in any case.

One thing for later is that for odin models we use the parameter setting to get information about the order of variables - we never added that into the first version and will have to add it in, but should not be much work.

File: inst/include/mode/r/helpers.hpp (hunk 1 above)
Comment (hillalex) [reply inst/include/mode/r/helpers.hpp:137]: Ok, makes sense about time resetting being key; but shouldn't we should reset state whenever time is reset, even if new pars not provided?

Comment (richfitz) [reply inst/include/mode/r/helpers.hpp:137]: Probably, yes! That's partly why I figured we could come back to this later :)

Comment (hillalex) [reply inst/include/mode/r/helpers.hpp:137]: Ok cool, shall we leave this as is?

Comment (richfitz) [reply inst/include/mode/r/helpers.hpp:137]: yup, made an issue for later: https://mrc-ide.myjetbrains.com/youtrack/issue/mrc-3139

**Review by richfitz — APPROVED** [review]

**Pull request merged**